File: 112.js
// LeetCode 112 Path Sum 
// References:https://leetcode.com/problems/path-sum/solutions/2432844/very-easy-100-3-line-explained-java-c-python-js-c-python3/?envType=study-plan-v2&envId=top-interview-150

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
var hasPathSum = function(root, targetSum) {
    // if the tree is empty, root is NULL, return false
    if (root == null) return false;

    // if only a single node and node value equals to targetSum
    if (root.val == targetSum && root.left == null && root.right == null)
        return true;

    // Call the same function recursively for left subtree and right subtree
    return hasPathSum(root.left, targetSum - root.val) || hasPathSum(root.right, targetSum - root.val);
};

File: 112.java
// LeetCode 112 Path Sum 
// References: https://leetcode.com/problems/path-sum/?envType=study-plan-v2&envId=top-interview-150

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean rootToLeafPathSum(TreeNode root, int targetSum, int sum){
        if (root == null) return false;

        if (root.left == null && root.right == null){
            sum += root.val;
            if (sum == targetSum)
                return true;
        }

        return rootToLeafPathSum(root.left, targetSum, sum + root.val) ||
            rootToLeafPathSum(root.right, targetSum, sum + root.val);
    }

    public boolean hasPathSum(TreeNode root, int targetSum) {
        int sum = 0;
        return rootToLeafPathSum(root, targetSum, sum);
    }
}